File: solutions/ruby/all-your-base/all_your_base.rb
class BaseConverter
  def self.convert(from_base, from_digits, to_base)
    raise ArgumentError.new "invalid input base" unless from_base >= 2
    raise ArgumentError.new "invalid output base" unless to_base >= 2

    # convert sequence of digits in input base to whole integer value
    value = from_digits.each.with_index(1).reduce(0) { |sum, (d, i)|
      raise ArgumentError.new "negative digits not allowed" unless d >= 0
      raise ArgumentError.new "digit out of range" unless d < from_base
      sum + d * from_base ** (from_digits.length - i)
    }

    # convert whole integer value to sequence of digits in output base
    to_digits = []
    until value < to_base
      to_digits.prepend value % to_base
      value = value / to_base
    end
    to_digits.prepend value
  end
end

File: solutions/javascript/all-your-base/all-your-base.js
export const convert = (fromDigits, fromBase, toBase) => {
  const errorWrongFormat = new Error("Input has wrong format");
  const errorInputBase = new Error("Wrong input base");
  const errorOutputBase = new Error("Wrong output base");

  if (fromBase < 2) throw errorInputBase;
  if (toBase < 2) throw errorOutputBase;
  if (!fromDigits.length) throw errorWrongFormat;
  if (!fromDigits[0] && fromDigits.length > 1) throw errorWrongFormat;

  // convert sequence of digits in input base to whole integer value
  const positionalValue = (acc, d) => {
    if (d < 0 || d >= fromBase) throw errorWrongFormat;
    acc *= fromBase;
    acc += d;
    return acc;
  };
  let value = fromDigits.reduce(positionalValue, 0);

  // convert whole integer value to sequence of digits in output base
  const toDigits = [];
  while (value >= toBase) {
    toDigits.unshift(value % toBase);
    value = Math.floor(value / toBase);
  }
  toDigits.unshift(value);

  return toDigits;
};
